src/Http/Cookie/CookieCollection.php

```diff
@@ -315,4 +361,81 @@ protected function parseSetCookieHeader($values)
 
         return $cookies;
     }
+
+    /**
+     * Remove expired cookies from the collection.
+     *
+     * @param string $host The host to check for expired cookies on.
+     * @param string $path The path to check for expired cookies on.
+     * @return void
+     */
+    private function removeExpiredCookies($host, $path)
+    {
+        $time = time();
+        $hostPattern = '/' . preg_quote($host, '/') . '$/';
+
+        foreach ($this->cookies as $i => $cookie) {
+            $expires = $cookie->getExpiry();
+            $expired = ($expires > 0 && $expires < $time);
+
+            $pathMatches = strpos($path, $cookie->getPath()) === 0;
+            $hostMatches = preg_match($hostPattern, $cookie->getDomain());
+            if ($pathMatches && $hostMatches && $expired) {
+                unset($this->cookies[$i]);
+            }
+        }
+    }
+
+    /**
+     * Store the cookies contained in a response
+     *
+     * This method operates on the collection in a mutable way for backwards
+     * compatibility reasons. This method should not be used and is only
+     * provided for backwards compatibility.
+     *
+     * @param \Cake\Http\Client\Response $response The response to read cookies from
+     * @param string $url The request URL used for default host/path values.
+     * @return void
+     * @deprecated 3.5.0 Will be removed in 4.0.0. Use `addFromResponse()` instead.
+     */
+    public function store(ClientResponse $response, $url)
+    {
+        $host = parse_url($url, PHP_URL_HOST);
+        $path = parse_url($url, PHP_URL_PATH);
+        $path = $path ?: '/';
+
+        $header = $response->getHeader('Set-Cookie');
+        $cookies = $this->parseSetCookieHeader($header);
+        $cookies = $this->setRequestDefaults($cookies, $host, $path);
+        foreach ($cookies as $cookie) {
+            $this->cookies[] = $cookie;
+        }
+        $this->removeExpiredCookies($host, $path);
+    }
+
+    /**
+     * Get all cookie data as arrays.
+     *
+     * This method should not be used and is only provided for backwards compatibility.
+     *
+     * @return array
+     * @deprecated 3.5.0 Will be removed in 4.0.0
```

Comment:
That is correct for this method. I want to replace the `Http\Client\CookieCollection` with this new class. In order to do that, this new class needs to provide the same public methods as the old class. However, I feel that the old methods are a bit clunky in both name and parameters, so I would like to deprecate them.

Another option I considered was to modify `Http\Client\CookieCollection` to extend this new class and leave the backwards compatibility there. I'm still open to that approach if people think it would be easier to understand in the future.